File: src/cfrpy/algorithms/algo.py
from __future__ import annotations

import structlog

logger = structlog.get_logger(__name__)

from typing import Protocol, Generic
from typing import Callable, Dict, Optional
from dataclasses import dataclass
import itertools
import pickle

from ..game import Game, A_inv, I


class Implementation(Protocol[A_inv, I]):
    @property
    def regrets(self) -> Dict[I, Dict[A_inv, float]]:
        ...

    @property
    def strategies(self) -> Dict[I, Dict[A_inv, float]]:
        ...

    @property
    def touched(self) -> int:
        # how many nodes touched
        ...

    def _run_iteration(self, game: Callable[[], Game[A_inv, I]]):
        ...


@dataclass
class Runner(Generic[A_inv, I]):
    impl: Implementation[A_inv, I]
    game: Callable[[], Game[A_inv, I]]

    pattern: str
    until: int
    checkpt: Optional[int] = None

    def run(self):
        impl, game = self.impl, self.game
        checkpt = self.checkpt

        for it in range(self.until):
            impl._run_iteration(game)

            if not it % 10:
                logger.info(
                    "iteration",
                    it=it,
                    touched=impl.touched,
                    infosets=len(impl.regrets),
                )

            if checkpt is not None and not (1 + it) % checkpt:
                logger.info(f"saving to {self.pattern.format((1 + it) // checkpt)}")
                self.save((1 + it) // checkpt)

    def save(self, it: int):
        path = self.pattern.format(it)
        with open(path, "wb") as f:
            progress = (self.impl.regrets, self.impl.strategies)
            pickle.dump(progress, f)

File: src/cfrpp/implementC/algo.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <math.h>

#include "logging.h"

#include "game.h"

#define MAX_REGRET_LENGTH 1000
#define MAX_STRATEGY_LENGTH 1000

struct Implementation {
    struct Dict *regrets;
    struct Dict *strategies;
    int touched;
    void (*_run_iteration)(struct Implementation *, Game (*)());
};

struct Runner {
    struct Implementation *impl;
    Game (*game)();
    char *pattern;
    int until;
    int checkpt;
    void (*run)(struct Runner *);
    void (*save)(struct Runner *, int);
};

void _run_iteration(struct Implementation *, Game (*)());
void run(struct Runner *);
void save(struct Runner *, int);

struct Implementation *create_Implementation() {
    struct Implementation *impl = (struct Implementation *) malloc(sizeof(struct Implementation));
    impl->regrets = create_Dict();
    impl->strategies = create_Dict();
    impl->touched = 0;
    impl->_run_iteration = _run_iteration;
    return impl;
}

typedef struct {
    int touched;
    float regrets[MAX_REGRET_LENGTH][MAX_REGRET_LENGTH];
    float strategies[MAX_STRATEGY_LENGTH][MAX_STRATEGY_LENGTH];
} Implementation;

typedef struct {
    Implementation *impl;
    Game *game;
    char *pattern;
    int until;
    int checkpt;
} Runner;

void run(Runner *runner) {
    Implementation *impl = runner->impl;
    Game *game = runner->game;
    int checkpt = runner->checkpt;

    for (int it = 0; it < runner->until; it++) {
        // impl->_run_iteration(game);

        if (it % 10 == 0) {
            printf("iteration %d, touched %d, infosets %d\n", it, impl->touched, MAX_REGRET_LENGTH);
        }

        if (checkpt != 0 && (1 + it) % checkpt == 0) {
            char fileName[100];
            sprintf(fileName, runner->pattern, (1 + it) / checkpt);
            printf("saving to %s\n", fileName);
            // self.save((1 + it) / checkpt);
        }
    }
}

void save(Runner *runner, int it) {
    char fileName[100];
    sprintf(fileName, runner->pattern, it);
    FILE *file = fopen(fileName, "wb");
    if (file == NULL) {
        printf("Error opening file!\n");
        exit(1);
    }
    // progress = (self.impl.regrets, self.impl.strategies)
    fwrite(runner->impl->regrets, sizeof(float), MAX_REGRET_LENGTH * MAX_REGRET_LENGTH, file);
    fwrite(runner->impl->strategies, sizeof(float), MAX_STRATEGY_LENGTH * MAX_STRATEGY_LENGTH, file);
    fclose(file);
}
